### PR TITLE
planner: refactor Converting Partition Keys for shuffle hash join

### PR DESCRIPTION
**Pull request by fzhedu**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tics/issues/1865 

Problem Summary:
we should enforce a projection for hash partition join only for the partition keys which are a subset of join keys, but the current code enforces a projection for all join keys.

### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed:

How it Works:

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Integration test

Side effects

### Release note 

- refactor converting partition keys for shuffle hash join

### CODE REVIEW

File: planner/core/task.go [hunk 1]
```diff
@@ -716,12 +714,8 @@ func (p *PhysicalHashJoin) convertJoinKeyForTiFlashIfNeed(lTask, rTask *mppTask)
 			rCast := expression.BuildCastFunction(p.ctx, rKey, cType)
 			rKey = appendExpr(rProj, rCast)
 		}
-		if lMask[i] || rMask[i] {
```
Comment (hanfei1991) [inline planner/core/task.go:719]: Why removes this check?

File: store/mockstore/unistore/cophandler/mpp_exec.go [hunk 2]
```diff
@@ -345,6 +345,25 @@ type joinExec struct {
 	inited       bool
 }
 
+func (e *joinExec) getHashKey(keyCol types.Datum, fieldType *types.FieldType) (str string, err error) {
```
Comment (hanfei1991) [inline store/mockstore/unistore/cophandler/mpp_exec.go:348]: Why should we do this convert?

File: planner/core/task.go (hunk 1 above)
Comment (fzhedu) [reply planner/core/task.go:719]: as HashJoin [ToPB](https://github.com/pingcap/tidb/blob/3f6dfe91daab7acb178ad24941dfca1745993afd/planner/core/plan_to_pb.go#L335) does not serialize the `p.EqualConditions`, it is enough that we only convert the data type of partition keys.

File: store/mockstore/unistore/cophandler/mpp_exec.go (hunk 2 above)
Comment (fzhedu) [reply store/mockstore/unistore/cophandler/mpp_exec.go:348]: if not, 1 != decimal(1.0)

Comment (hanfei1991) [reply store/mockstore/unistore/cophandler/mpp_exec.go:348]: Why not planner convert types in advance?

Comment (fzhedu) [reply store/mockstore/unistore/cophandler/mpp_exec.go]: updated, introduce a commen type to align join key types when get hash keys of string type.

**Review by hanfei1991 — APPROVED** [review]

**Review by rebelice — APPROVED** [review]
/lgtm

**Pull request merged**